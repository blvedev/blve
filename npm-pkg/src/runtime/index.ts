export type BlveComponent = {
  updatedFlag: boolean;
  valUpdateMap: number;
  blkRenderedMap: number;
  blkUpdateMap: number;
  internalElement: BlveInternalElement;
  currentVarBit: number;
  currentIfBlkBit: number;
  ifBlkRenderers: { [key: string]: () => void };
  isMounted: boolean;
  // componentElmentSetter: (innerHtml: string, topElmTag: string,topElmAttr: {[key: string]: string}) => void
  __blve_update: () => void;
  __blve_after_mount: () => void;
  // __blve_init: () => void;
  // __blve_destroy: () => void;
  // __blve_update_component: () => void;
  // __blve_update_component_end: () => void;
  // __blve_update_component_start: () => void;
  // __blve_update_end: () => void;
  // __blve_update_start: () => void;
  // __blve_init_component: () => void;
};

type BlveInternalElement = {
  innerHtml: string;
  topElmTag: string;
  topElmAttr: { [key: string]: string };
};

export const __BLVE_INIT_COMPONENT = function (this: BlveComponent) {
  this.updatedFlag = false;
  this.valUpdateMap = 0;
  this.blkRenderedMap = 0;
  this.blkUpdateMap = 0;
  this.currentVarBit = 0;
  this.currentIfBlkBit = 0;
  this.isMounted = false;

  const genBitOfVariables = function* (this: BlveComponent) {
    while (true) {
      if (this.currentVarBit === 0) {
        this.currentVarBit = 1;
        yield this.currentVarBit;
      } else {
        this.currentVarBit <<= 1;
        yield this.currentVarBit;
      }
    }
  }.bind(this);

  const genBitOfIfBlks = function* (this: BlveComponent) {
    while (true) {
      if (this.currentVarBit === 0) {
        this.currentVarBit = 1;
        yield this.currentVarBit;
      } else {
        this.currentVarBit <<= 1;
        yield this.currentVarBit;
      }
    }
  }.bind(this);

  const componentElementSetter = function (
    this: BlveComponent,
    innerHtml: string,
    topElmTag: string,
    topElmAttr: { [key: string]: string } = {}
  ) {
    this.internalElement = {
      innerHtml,
      topElmTag,
      topElmAttr,
    };
  }.bind(this);

  const setAfterMount = function (this: BlveComponent, afterMount: () => void) {
    this.__blve_after_mount = afterMount;
  }.bind(this);

  const mount = function (
    this: BlveComponent,
    elm: HTMLElement
  ): BlveComponent {
    if (this.isMounted) throw new Error("Component is already mounted");
    elm.innerHTML = `<${this.internalElement.topElmTag} ${Object.keys(
      this.internalElement.topElmAttr
    )
      .map((key) => `${key}="${this.internalElement.topElmAttr[key]}"`)
      .join(" ")}>${this.internalElement.innerHtml}</${
      this.internalElement.topElmTag
    }>`;
    this.__blve_after_mount();
    this.isMounted = true;
    return this;
  }.bind(this);

  const insertBefore = function (
    this: BlveComponent,
    elm: HTMLElement,
    anchor: HTMLElement | null
  ): BlveComponent {
    if (this.isMounted) throw new Error("Component is already mounted");
    const componentElm = createDomElementFromBlveElement(this.internalElement);
    elm.insertBefore(componentElm, anchor);
    this.__blve_after_mount();
    this.isMounted = true;
    return this;
  }.bind(this);

  const updateComponent = function (
    this: BlveComponent,
    updateFunc: () => void
  ) {
    this.__blve_update = (() => {
      if (!this.updatedFlag) return;
      updateFunc.call(this);
      this.updatedFlag = false;
      this.valUpdateMap = 0;
      this.blkUpdateMap = 0;
    }).bind(this);
  }.bind(this);

  const createReactive = function <T>(this: BlveComponent, v: T) {
    return new valueObj<T>(v, this, genBitOfVariables().next().value);
  }.bind(this);

  const createIfBlock = function (
    this: BlveComponent,
    name: string,
    blveElement: BlveInternalElement,
    parentElement: HTMLElement,
    refElement: HTMLElement | null,
    postRender: () => void
  ) {
    const ifBlkBit = genBitOfIfBlks().next().value;
    this.ifBlkRenderers[name] = (() => {
      const componentElm = createDomElementFromBlveElement(blveElement);
      parentElement.insertBefore(componentElm, refElement);
      postRender();
      (this.blkRenderedMap |= ifBlkBit), (this.blkUpdateMap |= ifBlkBit);
    }).bind(this);
  }.bind(this);

  const renderIfBlock = function (this: BlveComponent, name: string) {
    if (!this.ifBlkRenderers[name]) return;
    this.ifBlkRenderers[name]();
  };

  return {
    __BLVE_SET_COMPONENT_ELEMENT: componentElementSetter,
    __BLVE_UPDATE_COMPONENT: updateComponent,
    __BLVE_AFTER_MOUNT: setAfterMount,
    __BLVE_REACTIVE: createReactive,
    __BLVE_CREATE_IF_BLOCK: createIfBlock,
    __BLVE_RENDER_IF_BLOCK: renderIfBlock,
    __BLVE_COMPONENT_RETURN: {
      mount,
      insertBefore,
    },
  };
}.bind({} as BlveComponent);

class valueObj<T> {
  constructor(
    private _v: T,
    private componentObj: BlveComponent,
    private symbolIndex: number
  ) {}

  set v(v: T) {
    if (this._v === v) return;
    this._v = v;
    this.componentObj.valUpdateMap |= this.symbolIndex;
    if (!this.componentObj.updatedFlag) {
      Promise.resolve().then(this.componentObj.__blve_update.bind(this));
      this.componentObj.updatedFlag = true;
    }
  }

  get v() {
    return this._v;
  }
}

export function __BLVE_ESCAPE_HTML(text: any): string {
  const map: { [key: string]: string } = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&#039;",
  };

  return String(text).replace(/[&<>"']/g, function (m: string): string {
    return map[m];
  });
}

export function __BLVE_GET_ELM_REFS(ids: string[], preserveId: number) {
  return ids.map((id, index) => {
    const e = document.getElementById(id)!;
    (2 ** index) & preserveId && e.removeAttribute("id");
    return e;
  });
}

export function __BLVE_ADD_EV_LISTENER(
  elm: HTMLElement,
  evName: string,
  evFunc: EventListener
) {
  elm.addEventListener(evName, evFunc);
}

export function __BLVE_REPLACE_INNER_HTML(content: any, elm: Node) {
  elm.textContent = __BLVE_ESCAPE_HTML(content);
}

export function __BLVE_REPLACE_TEXT(content: any, elm: HTMLElement) {
  elm.innerHTML = __BLVE_ESCAPE_HTML(content);
}

export function __BLVE_REPLACE_ATTR(
  key: string,
  content: any,
  elm: HTMLElement
) {
  if (content === undefined && elm.hasAttribute(key)) {
    elm.removeAttribute(key);
    return;
  }
  (elm as any)[key] = String(content);
}

export function __BLVE_INSERT_EMPTY(
  parent: HTMLElement,
  anchor: HTMLElement | null
) {
  const empty = document.createTextNode(" ");
  parent.insertBefore(empty, anchor);
  return empty;
}

export function __BLVE_INSERT_CONTENT(
  content: string,
  parent: HTMLElement,
  anchor: HTMLElement | null
) {
  const contentNode = document.createTextNode(content);
  parent.insertBefore(contentNode, anchor);
  return contentNode;
}

export function __CREATE_BLVE_ELEMENT(
  innerHtml: string,
  topElmTag: string,
  topElmAttr: { [key: string]: string }
): BlveInternalElement {
  return {
    innerHtml,
    topElmTag,
    topElmAttr,
  };
}

export const createDomElementFromBlveElement = function (
  blveElement: BlveInternalElement
): HTMLElement {
  const componentElm = document.createElement(blveElement.topElmTag);
  Object.keys(blveElement.topElmAttr).forEach((key) => {
    componentElm.setAttribute(key, blveElement.topElmAttr[key]);
  });
  componentElm.innerHTML = blveElement.innerHtml;
  return componentElm;
};
