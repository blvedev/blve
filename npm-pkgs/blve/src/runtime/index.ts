export type ComponentDeclaration = (args?: {
  [key: string]: any;
}) => BlveModuleExports;

export type BlveModuleExports = {
  mount: (elm: HTMLElement) => BlveComponentState;
  insert: (elm: HTMLElement, anchor: HTMLElement | null) => BlveComponentState;
};

export type BlveComponentState = {
  updatedFlag: boolean;
  valUpdateMap: number;
  blkRenderedMap: number;
  blkUpdateMap: number;
  internalElement: BlveInternalElement;
  currentVarBit: number;
  currentIfBlkBit: number;
  ifBlkRenderers: { [key: string]: () => void };
  isMounted: boolean;
  // componentElmentSetter: (innerHtml: string, topElmTag: string,topElmAttr: {[key: string]: string}) => void
  __blve_update: () => void;
  __blve_after_mount: () => void;
  // __blve_init: () => void;
  // __blve_destroy: () => void;
  // __blve_update_component: () => void;
  // __blve_update_component_end: () => void;
  // __blve_update_component_start: () => void;
  // __blve_update_end: () => void;
  // __blve_update_start: () => void;
  // __blve_init_component: () => void;
};

type BlveInternalElement = {
  innerHtml: string;
  topElmTag: string;
  topElmAttr: { [key: string]: string };
};

export const $$blveInitComponent = function (this: BlveComponentState) {
  this.updatedFlag = false;
  this.valUpdateMap = 0;
  this.blkRenderedMap = 0;
  this.blkUpdateMap = 0;
  this.currentVarBit = 0;
  this.currentIfBlkBit = 0;
  this.isMounted = false;
  this.ifBlkRenderers = {};

  const genBitOfVariables = function* (this: BlveComponentState) {
    while (true) {
      if (this.currentVarBit === 0) {
        this.currentVarBit = 1;
        yield this.currentVarBit;
      } else {
        this.currentVarBit <<= 1;
        yield this.currentVarBit;
      }
    }
  }.bind(this);

  const genBitOfIfBlks = function* (this: BlveComponentState) {
    while (true) {
      if (this.currentIfBlkBit === 0) {
        this.currentIfBlkBit = 1;
        yield this.currentIfBlkBit;
      } else {
        this.currentIfBlkBit <<= 1;
        yield this.currentIfBlkBit;
      }
    }
  }.bind(this);

  const componentElementSetter = function (
    this: BlveComponentState,
    innerHtml: string,
    topElmTag: string,
    topElmAttr: { [key: string]: string } = {}
  ) {
    this.internalElement = {
      innerHtml,
      topElmTag,
      topElmAttr,
    };
  }.bind(this);

  const setAfterMount = function (
    this: BlveComponentState,
    afterMount: () => void
  ) {
    this.__blve_after_mount = afterMount;
  }.bind(this);

  const mount = function (
    this: BlveComponentState,
    elm: HTMLElement
  ): BlveComponentState {
    if (this.isMounted) throw new Error("Component is already mounted");
    elm.innerHTML = `<${this.internalElement.topElmTag} ${Object.keys(
      this.internalElement.topElmAttr
    )
      .map((key) => `${key}="${this.internalElement.topElmAttr[key]}"`)
      .join(" ")}>${this.internalElement.innerHtml}</${
      this.internalElement.topElmTag
    }>`;
    this.__blve_after_mount();
    this.isMounted = true;
    return this;
  }.bind(this);

  const insert = function (
    this: BlveComponentState,
    elm: HTMLElement,
    anchor: HTMLElement | null
  ): BlveComponentState {
    if (this.isMounted) throw new Error("Component is already mounted");
    const componentElm = createDomElementFromBlveElement(this.internalElement);
    elm.insertBefore(componentElm, anchor);
    this.__blve_after_mount();
    this.isMounted = true;
    return this;
  }.bind(this);

  const updateComponent = function (
    this: BlveComponentState,
    updateFunc: () => void
  ) {
    this.__blve_update = (() => {
      if (!this.updatedFlag) return;
      updateFunc.call(this);
      this.updatedFlag = false;
      this.valUpdateMap = 0;
      this.blkUpdateMap = 0;
    }).bind(this);
  }.bind(this);

  const createReactive = function <T>(this: BlveComponentState, v: T) {
    return new valueObj<T>(v, this, genBitOfVariables().next().value);
  }.bind(this);

  const createIfBlock = function (
    this: BlveComponentState,
    name: string,
    blveElement: () => BlveInternalElement,
    getParentAndRefElement: () => [HTMLElement, HTMLElement | null],
    postRender: () => void
  ) {
    const ifBlkBit = genBitOfIfBlks().next().value;
    this.ifBlkRenderers[name] = (() => {
      const componentElm = createDomElementFromBlveElement(blveElement());
      const [parentElement, refElement] = getParentAndRefElement();
      parentElement.insertBefore(componentElm, refElement);
      postRender();
      (this.blkRenderedMap |= ifBlkBit), (this.blkUpdateMap |= ifBlkBit);
    }).bind(this);
  }.bind(this);

  const renderIfBlock = function (this: BlveComponentState, name: string) {
    if (!this.ifBlkRenderers[name]) return;
    this.ifBlkRenderers[name]();
  }.bind(this);

  return {
    $$blveSetComponentElement: componentElementSetter,
    $$blveUpdateComponent: updateComponent,
    $$blveAfterMount: setAfterMount,
    $$blveReactive: createReactive,
    $$blveCreateIfBlock: createIfBlock,
    $$blveRenderIfBlock: renderIfBlock,
    $$blveComponentReturn: {
      mount,
      insert,
    } as BlveModuleExports,
  };
};

class valueObj<T> {
  constructor(
    private _v: T,
    private componentObj: BlveComponentState,
    private symbolIndex: number
  ) {}

  set v(v: T) {
    if (this._v === v) return;
    this._v = v;
    this.componentObj.valUpdateMap |= this.symbolIndex;
    if (!this.componentObj.updatedFlag) {
      Promise.resolve().then(this.componentObj.__blve_update.bind(this));
      this.componentObj.updatedFlag = true;
    }
  }

  get v() {
    return this._v;
  }
}

export function $$blveEscapeHtml(text: any): string {
  const map: { [key: string]: string } = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&#039;",
  };

  return String(text).replace(/[&<>"']/g, function (m: string): string {
    return map[m];
  });
}

export function $$blveGetElmRefs(ids: string[], preserveId: number) {
  return ids.map((id, index) => {
    const e = document.getElementById(id)!;
    (2 ** index) & preserveId && e.removeAttribute("id");
    return e;
  });
}

export function $$blveAddEvListener(
  elm: HTMLElement,
  evName: string,
  evFunc: EventListener
) {
  elm.addEventListener(evName, evFunc);
}

export function $$blveReplaceText(content: any, elm: Node) {
  elm.textContent = $$blveEscapeHtml(content);
}

export function $$blveReplaceInnerHtml(content: any, elm: HTMLElement) {
  elm.innerHTML = $$blveEscapeHtml(content);
}

export function $$blveReplaceAttr(key: string, content: any, elm: HTMLElement) {
  if (content === undefined && elm.hasAttribute(key)) {
    elm.removeAttribute(key);
    return;
  }
  (elm as any)[key] = String(content);
}

export function $$blveInsertEmpty(
  parent: HTMLElement,
  anchor: HTMLElement | null
) {
  const empty = document.createTextNode(" ");
  parent.insertBefore(empty, anchor);
  return empty;
}

export function $$blveInsertContent(
  content: string,
  parent: HTMLElement,
  anchor: HTMLElement | null
) {
  const contentNode = document.createTextNode(content);
  parent.insertBefore(contentNode, anchor);
  return contentNode;
}

export function $$createBlveElement(
  innerHtml: string,
  topElmTag: string,
  topElmAttr: { [key: string]: string }
): BlveInternalElement {
  return {
    innerHtml,
    topElmTag,
    topElmAttr,
  };
}

export const createDomElementFromBlveElement = function (
  blveElement: BlveInternalElement
): HTMLElement {
  const componentElm = document.createElement(blveElement.topElmTag);
  Object.keys(blveElement.topElmAttr).forEach((key) => {
    componentElm.setAttribute(key, blveElement.topElmAttr[key]);
  });
  componentElm.innerHTML = blveElement.innerHtml;
  return componentElm;
};
