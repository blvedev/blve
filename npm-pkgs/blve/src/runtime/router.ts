import { BlveModuleExports, ComponentDeclaration } from ".";
import { routes as autoRoutes } from "blve-auto-routing-plugin";

export type ComponentLoader = () => Promise<ComponentDeclaration>;

export type Route = {
  path: string;
  component: ComponentLoader;
};

export class Router {
  routes: Route[];
  notFound: () => void;
  currentComponent: BlveModuleExports | null;
  renderingTarget!: {
    parent: HTMLElement;
    anchor: HTMLElement | null;
  };

  constructor() {
    this.routes = [];
    this.notFound = () => {};
    this.currentComponent = null;
    window.addEventListener("popstate", this.handlePopState.bind(this));
  }

  addRoute(path: string, componentLoader: ComponentLoader) {
    this.routes.push({ path, component: componentLoader });
  }

  setNotFound(notFoundHandler: () => void) {
    this.notFound = notFoundHandler;
  }

  navigate(path: string) {
    window.history.pushState({}, path, window.location.origin + path);
    this.handleRoute(path);
  }

  async handleRoute(path: string) {
    const route = this.routes.find((route) => route.path === path);
    if (route) {
      const component = await route.component();
      this.renderComponent(component);
    } else {
      this.notFound();
    }
  }

  handlePopState() {
    this.handleRoute(window.location.pathname);
  }

  renderComponent(component: ComponentDeclaration) {
    // TODO: Execute destroy function of the current component
    // if (this.currentComponent && this.currentComponent.__blve_destroy) {
    //   this.currentComponent.__blve_destroy();
    // }
    this.currentComponent = component();
    this.currentComponent.mount(document.getElementById("app")!);
  }

  initialize(parent: HTMLElement, anchor: HTMLElement | null) {
    this.renderingTarget = { parent, anchor };
    this.handleRoute(window.location.pathname);
  }

  // DO NOT EXECUTE THIS FUNCTION IN NON-VITE ENVIRONMENTS
  async initializeWithAutoRoutes(
    parent: HTMLElement,
    anchor: HTMLElement | null
  ) {
    (autoRoutes as Route[]).forEach((route) => {
      this.addRoute(route.path, route.component);
    });
    this.initialize(parent, anchor);
  }
}

export const router = new Router();
