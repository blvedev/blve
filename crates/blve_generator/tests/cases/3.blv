html:
  <input @input="onInput" :value="balue" />
  <div>${balue}</div>
  <button @click="xxx">xxx</button>
script:
  const balue = ""
  function onInput(e) {
    balue = e.target.value
  }
  function xxx() {
    balue += "xxx"
  }
