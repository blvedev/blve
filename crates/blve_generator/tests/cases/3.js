
import { reactiveValue,getElmRefs,addEvListener,genUpdateFunc,escapeHtml,replaceText,replaceAttr,insertEmpty } from 'blve/dist/runtime'
export default function(elm) {
    const refs = [0, false, null];
    const balue = reactiveValue("", 1, refs)
    function onInput(e) {
      balue.v = e.target.value
    }
    function xxx() {
      balue.v += "xxx"
    }

    elm.innerHTML = `<div><input id="mSZVkKzedMphuVkGNIHeF" value="${balue.v}" /><div id="tzPQlyiTuzVvo_BLFwCqV">${escapeHtml(balue.v)}</div><button id="syHUaALEsaXMHcAsLMvyE">xxx</button></div>`;

    const [mSZVkKzedMphuVkGNIHeFRef,tzPQlyiTuzVvo_BLFwCqVRef,syHUaALEsaXMHcAsLMvyERef] = getElmRefs(["mSZVkKzedMphuVkGNIHeF","tzPQlyiTuzVvo_BLFwCqV","syHUaALEsaXMHcAsLMvyE"], 7);

    addEvListener(mSZVkKzedMphuVkGNIHeFRef, "input", onInput);

    addEvListener(syHUaALEsaXMHcAsLMvyERef, "click", xxx);

    refs[2] = genUpdateFunc(() => {
        refs[0]  & 1 && replaceAttr("value", balue.v, mSZVkKzedMphuVkGNIHeFRef);

        refs[0] & 1 && replaceText(`${escapeHtml(balue.v)}`, tzPQlyiTuzVvo_BLFwCqVRef);

    });

}