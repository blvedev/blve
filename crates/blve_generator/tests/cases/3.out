
import { reactiveValue,getElmRefs,addEvListener,genUpdateFunc,escapeHtml,replaceText,replaceAttr } from 'blve/dist/runtime'
export default function(elm) {
    const refs = [0, false, null];
    const balue = reactiveValue("", 1, refs)
    function onInput(e) {
      balue.v = e.target.value
    }
    function xxx() {
      balue.v += "xxx"
    }

    elm.innerHTML = `<input id="test" value="${balue.v}"/><div id="test">${escapeHtml(balue.v)}</div><button id="test">xxx</button>`;

    const [testRef,testRef,testRef] = getElmRefs(["test","test","test"], 7);

    addEvListener(testRef, "input", onInput);

    addEvListener(testRef, "click", xxx);

    refs[2] = genUpdateFunc(() => {
        refs[0]  & 1 && replaceAttr("value", balue.v, testRef);
    
        refs[0] & 1 && replaceText(`${escapeHtml(balue.v)}`, testRef);
    
    });

}